### PR TITLE
Feat/e2e testing for article sorting

### PR DESCRIPTION
**Pull request by John-Paul-Larkin**
# ✨ Codu Pull Request 💻

## Pull Request details

- Add Playwright tests for sorting on the article page
- Run test for three sorting modes -  "Newest", "Oldest" and "Top"

- Like count added to ArticlePreview
- I was required to pass the like count to the component, so If it non-zero, I display it on the Article preview with the heart icon. 
- I also added the count as a data attribute, so we can easily remove this if you dont like how it looks.

Article preview with like count

## Any Breaking changes

- None

### CODE REVIEW

File: components/ArticlePreview/ArticlePreview.tsx [hunk 1]
```diff
@@ -124,6 +127,15 @@ const ArticlePreview: NextPage<Props> = ({
                 <>
                   <span aria-hidden="true">&middot;</span>
                   <span>{readTime} min read</span>
+                  {likes && (
```
Comment (JohnAllenTech) [inline components/ArticlePreview/ArticlePreview.tsx:131]: Oh interesting because you need a way to know how many likes an article has without clicking into it. That makes sense

File: e2e/articles.spec.ts [hunk 2]
```diff
@@ -220,4 +220,52 @@ test.describe("Authenticated Articles Page", () => {
 
     await expect(page.getByText(commentContent)).toBeVisible();
   });
+
+  test.describe("Article sort order on article page", () => {
+    const sortOptions = ["newest", "oldest", "top"];
+
+    for (const sortOption of sortOptions) {
```
Comment (JohnAllenTech) [inline e2e/articles.spec.ts]: Id recommend breaking ["newest", "oldest", "top"] into their own tests.

Yes we are repeating outselves but we also get test isolation ie. currently if this test breaks we will need to go into the actual test to see which option it breaks on

File: e2e/articles.spec.ts [hunk 3]
```diff
@@ -220,4 +220,52 @@ test.describe("Authenticated Articles Page", () => {
...
+            date: article.querySelector("time")?.dateTime,
+            likes: parseInt(
+              article
+                .querySelector("[data-likes]")
+                ?.getAttribute("data-likes") || "0",
+            ),
+          }));
+        });
+
+        if (sortOption === "newest") {
```
Comment (JohnAllenTech) [inline e2e/articles.spec.ts]: Comment above would clean up all this.

**Review by JohnAllenTech — COMMENTED** [review]
Just a few comments. Overall it LGTM

File: e2e/articles.spec.ts (hunk 2 above)
Comment (John-Paul-Larkin) [reply e2e/articles.spec.ts]: Yeah, I was not sure what to do there. I will split it into separate tests. Thank John.

BTW is using the data attribute(for the like count ) the correct way to go about this?

Comment (JohnAllenTech) [reply e2e/articles.spec.ts]: If your free to sync for 5 mins I can show you the way I get the most specific way to target an element.

TLDR if your not free. 

- Install playwright plugin for vs code
- Open the plugin view in vs code
- Use the pick locator function and hover over the element you want to target
- The plugin will give you the best locator to use

Comment (John-Paul-Larkin) [reply e2e/articles.spec.ts]: @JohnAllenTech 
Great, I just hacked my way through writing a test today, without reading the docs 🫢 
I just downloaded that plugin, so will give it a try now. 

That would be greta. I am online if you are free?

**Comment by NiallJoeMaher** [comment]
@JohnAllenTech, this one good to go?

File: e2e/articles.spec.ts [hunk 4]
```diff
@@ -220,4 +220,58 @@ test.describe("Authenticated Articles Page", () => {
 
     await expect(page.getByText(commentContent)).toBeVisible();
   });
+
+  test("Should sort articles by Newest", async ({ page }) => {
```
Comment (JohnAllenTech) [inline e2e/articles.spec.ts]: Do we want to copy and paste these tests into the unauthenticated view?

Does it need to be in both or just unauthenticated?

Given I believe the functionality of sorting is unaffected by a users auth status I think it only needs to be in unauthenticated

**Review by JohnAllenTech — COMMENTED** [review]
Just a small comment. Tests LGTM

File: e2e/articles.spec.ts (hunk 4 above)
Comment (John-Paul-Larkin) [reply e2e/articles.spec.ts]: hm... Yes, I guess it shouldnt matter really?
I will copy it to unauthenticated.

File: components/ArticlePreview/ArticlePreview.tsx (hunk 1 above)
Comment (NiallJoeMaher) [reply components/ArticlePreview/ArticlePreview.tsx:131]: Why do we need to know the amount of likes before it's loaded? 🤔 I don't think it's shown on the card?

Comment (John-Paul-Larkin) [reply components/ArticlePreview/ArticlePreview.tsx:131]: @NiallJoeMaher 
> Why do we need to know the amount of likes before it's loaded? 🤔 I don't think it's shown on the card?

I added it to the card in this PR. 
if you take a look at the PR description you will see how it looks. 
Heart icon with the like count. If there are no likes it wont show.

I figured you wouldnt like this 🤣  so I added it as a data attribute and can edit to remove it from the card.
The like count has to be included somewhere in the html, so playwright can grab it. For the sort by top test.

**Review by NiallJoeMaher — APPROVED** [review]
